### PR TITLE
Feature train no added noise

### PR DESCRIPTION
**Pull request by tara-leininger**
This version can run and 'un-normalizes' the bin weights for the plots, but not the actual data yet. Can use multiple files from batch submission

### CODE REVIEW

File: config_multi.py [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+num_files = 2
+
+def get_files(filetype):
+    if filetype == 'Fuzz':
+        list_of_names = []
+        for name in glob.glob('ntup_photon_energy'+str(energy)+'*Prod*_n'+str(num_events)+'.root'):
+            list_of_names.append(name)
+        list_of_names.sort()
+        list_of_names = list_of_names[:num_files]
+        " ".join(list_of_names)
```
Comment (kpedro88) [inline config_multi.py:15]: better to keep the list as a list

File: dataset.py [hunk 2]
```diff
@@ -7,13 +7,24 @@
 import torch 
 import math
 
-def get_all_histograms(file_path):
-    file = up.open(file_path)
-    tree = file["g4SimHits/tree"]
-    branch = tree["bin_weights"].array()
+def get_branch(file_paths):
+    file_path = file_paths.split(' ')
```
Comment (kpedro88) [inline dataset.py:11]: if it's an actual list (see above), can avoid need for split here

File: dataset.py [hunk 3]
```diff
@@ -39,28 +47,42 @@ def add_noise(data, sigma):
...
-    def __init__(self, root_file, sigma):
-        self.root_file = root_file
-        self.sigma = sigma
-        self.histograms = get_all_histograms(root_file)
-
+    allowed_transforms = ["none","normalize","log10"]
+    def __init__(self, fuzzy_root, sharp_root, transform='none'):
+        self.sharp_root = sharp_root
+        self.fuzzy_root = fuzzy_root
+        #self.sigma = sigma
```
Comment (kpedro88) [inline dataset.py:54]: delete

File: dataset.py [hunk 4]
```diff
@@ -39,28 +47,42 @@ def add_noise(data, sigma):
...
+        self.transform = transform
+        if self.transform not in self.allowed_transforms:
+            raise ValueError("Unknown transform: {}".format(self.transform))
+        
     def __len__(self):
-        return len(self.histograms)
+        if len(self.sharp_branch) == len(self.fuzzy_branch):
+            return len(self.sharp_branch)
+        else:
+            print("Sharp and fuzzy dataset lengths do not match")
```
Comment (kpedro88) [inline dataset.py:65]: better to stop execution in this case: `raise RuntimeError(...)`

File: train.py [hunk 5]
```diff
@@ -33,14 +34,18 @@
 parser.add_argument("--outf", type=str, default="logs", help='Name of folder to be used to store outputs')
 parser.add_argument("--epochs", type=int, default=50, help="Number of training epochs")
 parser.add_argument("--lr", type=float, default=1e-3, help="Initial learning rate")
-parser.add_argument("--trainfile", type=str, default="test.root", help='Path to .root file for training')
-parser.add_argument("--valfile", type=str, default="test.root", help='Path to .root file for validation')
+parser.add_argument("--trainfileSharp", type=str, default="test.root", help='Path to higher quality .root file for training')
+parser.add_argument("--trainfileFuzz", type=str, default="test.root", help='Path to lower quality .root file for training')
+parser.add_argument("--valfileSharp", type=str, default="test.root", help='Path to higher quality .root file for validation')
+parser.add_argument("--valfileFuzz", type=str, default="test.root", help='Path to lower quality .root file for validation')
```
Comment (kpedro88) [inline train.py:40]: these should have `nargs='+'` ([docs](https://docs.python.org/3/library/argparse.html#nargs)), then multiple values can be given on the command line separated by spaces (without need to quote or split into a list manually - argparse will do it automatically)

since "test.root" is not a real file, just remove the default value and add `required=True`

File: train.py [hunk 6]
```diff
@@ -50,47 +55,118 @@ def write_info_file():
...
+#assumes that the xmin, xmax, ymin, and ymax values are the same 
+#for all four high and low quality training and validation files
+file = up.open(args.valfileSharp)
+tree = file["g4SimHits/tree"]
+x_min = tree["xmin"].array().to_numpy()[0]
+x_max = tree["xmax"].array().to_numpy()[0]
+y_min = tree["ymin"].array().to_numpy()[0]
+y_max = tree["ymax"].array().to_numpy()[0]
+x_bins = tree["xbins"].array().to_numpy()[0]
+y_bins = tree["ybins"].array().to_numpy()[0]
```
Comment (kpedro88) [inline train.py:108]: this should go inside the `make_plots()` function below

File: train.py [hunk 7]
```diff
@@ -50,47 +55,118 @@ def write_info_file():
...
+        event_number = int(event_number)
+        means = np.mean(data[0][0].numpy())
+        stdevs = np.std(data[0][0].numpy())
+        
+        for i, elem in enumerate(binarray):
+            binarray[i] *= stdevs
+            binarray[i] += means
+        
+    else:
+        print('not output')
```
Comment (kpedro88) [inline train.py:130]: should this be an exception?

File: train.py [hunk 8]
```diff
@@ -50,47 +55,118 @@ def write_info_file():
...
+            binarray[i] += means
+        
+    else:
+        print('not output') 
+        
+    #builds axes for histogram given min/max and bin number
+    x_axis = []
+    count = 0
+    x_start = x_min
+    while count < 100:
```
Comment (kpedro88) [inline train.py:136]: should these 100s be x_bins and y_bins?

File: train.py [hunk 9]
```diff
@@ -50,47 +55,118 @@ def write_info_file():
...
+    fin = str(fin).replace(args.outf+'/samples/','')
+    plt.savefig(args.outf+'/plots/' + str(fin).replace( '.txt', '.png'))
+
 def init_weights(m):
     if type(m) == nn.Linear:
         torch.nn.init.xavier_uniform(m.weight)
         m.bias.data.fill_(0.01)
 
 def main():
+    random.seed(args.randomseed)
```
Comment (kpedro88) [inline train.py:167]: you should add `torch.manual_seed(args.randomseed)` to make the training reproducible

File: train.py [hunk 10]
```diff
@@ -171,13 +250,31 @@ def main():
...
     plt.savefig(args.outf + "/loss_plot.png")
+    
+    #write out training and validataion loss values to text files
+    tfileout = open("training_losses.txt","w")
+    vfileout = open("validation_losses.txt","w")
+    for i, elem in enumerate(training_losses):
+        tfileout.write("%f " % training_losses[i] + "\n")
+    for elem in enumerate(validation_losses):
+        tfileout.write("%f " % validation_losses[i] + "\n")
+        vfileout.write("%f " % validation_losses[i] + "\n")
```
Comment (kpedro88) [inline train.py:267]: I'm confused about what these loops are supposed to be doing.

The first loop could just be `for elem in training_losses`, since `training_losses[i]` is `elem`.

The second loop puts the entire output of `enumerate(validation_losses)` in `elem`, then uses `i` instead; but it doesn't define `i`, so it's just reusing the last value of `i` from the previous loop. It's also writing the validation loss into both the validation and training output files.

File: train.py [hunk 11]
```diff
@@ -171,13 +250,31 @@ def main():
...
+
+    random.seed(args.randomseed)
```
Comment (kpedro88) [inline train.py:269]: is this to make sure that you get the same rotations used in training for the sample images? if so, it would be good to add a comment to that effect

**Comment by kpedro88** [comment]
please undo the deletion of the .gitignore files in https://github.com/kpedro88/SimDenoising_training/pull/2/commits/e9214860ce3149dfa95c569eba6cf699b035da1e (also, why is this commit message train.py when the commit doesn't touch train.py?)

File: config_multi.py (hunk 1 above)
Comment (tara-leininger) [reply config_multi.py:15]: definitely agree that this could be better, but given that this works at the moment do you want me to table this for now?

File: train.py (hunk 5 above)
Comment (tara-leininger) [reply train.py:40]: same question as above^^

File: train.py (hunk 6 above)
Comment (tara-leininger) [reply train.py:97]: Will leave this for now given our discussion today

File: train.py (hunk 7 above)
Comment (tara-leininger) [reply train.py]: This part of the code just got cut in today's update, but that was initially just something that got left over from my working version that I accidentally left in there

File: train.py (hunk 8 above)
Comment (tara-leininger) [reply train.py:111]:
```suggestion
        x_axis = []
    count = 0
    x_start = x_min
    while count < y_bins:
        for i in range(x_bins):
            x = x_start + ((x_max-x_min)/x_bins)*float(i)
            x_axis.append(x)
        count = count + 1

    y_axis = []
    y_start = y_min
    for i in range(y_bins):
        y = y_start + ((y_max-y_min)/y_bins)*float(i)
        count = 0
        while count < x_bins:
            y_axis.append(y)
            count = count + 1
```

yes definitely. Also I know this is simple but I'm getting a little turned around with where each should go... also there's something similar going on  here: https://github.com/kpedro88/SimDenoising_training/blob/00b63b75282fb7c840e75c9addba6c4a24768046/dataset.py#L28-L35
so i should probably make a function so that doesn't get mixed up... just let me know if there's any reason that needs to stay separate

File: train.py (hunk 11 above)
Comment (tara-leininger) [reply train.py]: I moved this here:
https://github.com/kpedro88/SimDenoising_training/blob/28eb0d88dbac0548d9e2d5de2f0df70e4abca1a4/train.py#L70-L74
I think the placement makes it seem a little less out of nowhere but I've added a comment as well

File: train.py (hunk 10 above)
Comment (tara-leininger) [reply train.py:242]: yes this is definitely a mess... I think we had wanted to look at training and validation losses but I didn't finish this so I'll clean that up... 
```suggestion
    for elem in enumerate(training_losses):
        tfileout.write("%f " % elem + "\n")
    for elem in enumerate(validation_losses):
        vfileout.write("%f " % elem + "\n")
```

File: train.py (hunk 9 above)
Comment (tara-leininger) [reply train.py:142]:
```suggestion
    random.seed(args.randomseed)
    torch.manual_seed(args.randomseed)
```
I'm not entirely sure what randomness we are seeding with this (initial values for output data before going through CNN)?

File: config_multi.py (hunk 1 above)
Comment (kpedro88) [reply config_multi.py:15]: I suppose (though doing this by hand instead of using argparse's built-in features makes me sad)

File: train.py (hunk 9 above)
Comment (kpedro88) [reply train.py:142]: the torch random seed controls the initial values of the weights in the CNN before training (among other things)
